blocks/alternative_mutators/plus_minus/src/blocks.js

```diff
@@ -0,0 +1,1040 @@
+   * @param {!Array<string>} names An array of argument names to display.
+   * @param {!Array<string>} varIds An array of variable IDs associated with
+   *     those names.
+   * @this Blockly.Block
+   * @private
+   */
+  updateShape_: function(names, varIds) {
+    // In this case it is easiest to just reset and build from scratch.
+
+    // We need to remove args in reverse order so that it doesn't mess up
+    // as removeArg_ modifies our arrays.
+    for (var i = this.argIds_.length - 1; i >= 0; i--) {
+      this.removeArg_(this.argIds_[i]);
+    }
+
+    this.arguments_ = [];
+    this.varIds_ = [];
+    this.argumentVarModels_ = [];
+    this.argIds_ = [];
+
+    var length = varIds.length;
+    for (i = 0; i < length; i++) {
+      this.addArg_(names[i], varIds[i]);
+    }
+
+    Blockly.Procedures.mutateCallers(this);
+  },
+
+  /**
+   * Callback for the plus image. Adds an argument to the block and mutates
+   * callers to match.
+   */
+  plus: function() {
+    this.addArg_();
+    Blockly.Procedures.mutateCallers(this);
+  },
+
+  /**
+   * Callback for the minus image. Removes the argument associated with the
+   * given argument ID and mutates the callers to match.
+   * @this Blockly.Block
+   * @param argId
+   */
+  minus: function(argId) {
+    if (!this.argIds_.length) {
+      return;
+    }
+    this.removeArg_(argId);
+    Blockly.Procedures.mutateCallers(this);
+  },
+
+  /**
+   * Adds an argument to the block and updates the block's parallel tracking
+   * tracking arrays as appropriate.
+   * @param {string} opt_name An optional name for the argument.
+   * @param {string} opt_varId An optional variable ID for the argument.
+   * @this Blockly.Block
+   * @private
+   */
+  addArg_: function(opt_name, opt_varId) {
+    if (!this.arguments_.length) {
+      var withField = new Blockly.FieldLabel(
+          Blockly.Msg['PROCEDURES_BEFORE_PARAMS']);
+      this.getInput('TOP')
+          .appendField(withField, 'WITH');
+    }
+
+    var name = opt_name || Blockly.Variables.generateUniqueNameFromOptions(
+        Blockly.Procedures.DEFAULT_ARG, this.arguments_);
+    var variable = Blockly.Variables.getOrCreateVariablePackage(
+        this.workspace, opt_varId, name, '');
+    var argId = Blockly.utils.genUid();
+
+    this.addVarInput_(name, argId);
+    if (this.getInput('STACK')) {
+      this.moveInputBefore(argId, 'STACK');
+    } else {
+      this.moveInputBefore(argId, 'RETURN');
+    }
+
+    this.arguments_.push(name);
+    this.varIds_.push(variable.getId());
+    this.argumentVarModels_.push(variable);
+    this.argIds_.push(argId);
+  },
+
+  /**
+   * Removes the argument associated with the given argument ID from the block.
+   * @param {string} argId An ID used to track arguments on the block.
+   * @private
+   */
+  removeArg_: function(argId) {
+    if (!this.getInput(argId)) {
+      return;
+    }
+    // TODO: My life would be so much happier if this returned a boolean instead
```

Comment:
🤔

Reply:
Here's the offending implementation:
```
/**
 * Remove an input from this block.
 * @param {string} name The name of the input.
 * @param {boolean=} opt_quiet True to prevent error if input is not present.
 * @throws {Error} if the input is not present and opt_quiet is not true.
 */
Blockly.Block.prototype.removeInput = function(name, opt_quiet) {
  for (var i = 0, input; (input = this.inputList[i]); i++) {
    if (input.name == name) {
      if (input.type == Blockly.NEXT_STATEMENT) {
        this.statementInputCount--;
      }
      input.dispose();
      this.inputList.splice(i, 1);
      return;
    }
  }
  if (!opt_quiet) {
    throw Error('Input not found: ' + name);
  }
};
```